### PR TITLE
[KernelGen][Nvidia] Add _cdist_forward operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `_cdist_forward`. Computes the batched pairwise distance (cdist) between two collections of row vectors,
supporting p=2.0 (L2 norm) with a Metax backend specialization.

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_cdist_forward.py --level core` (NVIDIA H20)

### cdist_forward

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float32 | [4, 8] | 0.006160 | 0.005456 | 1.129 |
| float32 | [8, 16] | 0.006208 | 0.005952 | 1.043 |
| float32 | [16, 32] | 0.006208 | 0.006112 | 1.016 |
| float32 | [32, 64] | 0.032928 | 0.006240 | 5.277 |
| float32 | [64, 128] | 0.034176 | 0.009184 | 3.721 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| cdist_forward | **2.437** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (5 cases) | 2.437 | Primary |
| Tianshu | PASS | 0.991 | — |
| Muxi | FAIL | — | Error: "cdist_cuda" not implemented for 'Half'; Error: module 'flag_gems' has no attribute '_cdist_forward' |
| Ascend | PASS | — | — |
| Hygon | FAIL | 0.953 | Error: tuple index out of range |

## Files Changed
- `src/flag_gems/ops/_cdist_forward.py`: Triton kernel implementation
- `tests/test_cdist_forward.py`: Accuracy test
- `benchmark/test_cdist_forward.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.4)

### CODE REVIEW

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## Code Review: _cdist_forward Operator (Batched Pairwise Distance)

### 📋 Checklist Review

| # | Item | Status | Notes |
|---|------|--------|-------|
| 1 | CI Status | ✅ PASS | license/cla, triage passed |
| 2 | Merge Conflicts | ⚠️ CHECK | Verify mergeable state |
| 3 | Test Coverage | ✅ Good | Has `test_cdist_forward.py` |
| 4 | Benchmark Coverage | ✅ Good | Has benchmark with speedup metrics |
| 5 | Config Sort Order | ✅ Sorted | `_cdist_forward` in correct position |
| 6 | Kernel Registration | ✅ Correct | Registered in all `__init__.py` files |
| 7 | Missing Ops Check | ✅ N/A | No missing ops detected |
| 8 | NPU/CUDA Compatibility | ⚠️ Mixed | Metax backend specialized, Half dtype issue |
| 9 | Kernel Name vs Mark Name | ✅ Consistent | `_cdist_forward` used consistently |
| 10 | Logger Prefix | ✅ Good | Uses `flag_gems._cdist_forward` |
| 11 | Python Reserved Keywords | ✅ Good | No reserved keywords used |
| 12 | Code Quality | ✅ Good | Well-documented with comments |

### ✅ Strengths

1. **Complete Implementation Stack**
   - Main kernel: `src/flag_gems/ops/_cdist_forward.py`
   - Metax backend: `src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py`
   - Test file: `tests/test_cdist_forward.py`
   - Benchmark file: `benchmark/test_cdist_forward.py`
   - All configuration files updated

2. **Optimized Metax Backend**
   - 193 lines of specialized kernel for Metax hardware
   - Uses tiling strategy for efficient memory access
   - Achieves 2.437x speedup on Nvidia H20

3. **Good Algorithm Design**
   ```python
   # cdist(x1, x2, p=2) computes pairwise L2 distances
   # ||x1[i] - x2[j]||_2 for all i, j
   ```
   - Correctly implements L2 norm (p=2.0)
   - Handles batched inputs with broadcasting

4. **Comprehensive Tests**
   - Tests different shapes and dtypes
   - Compares with PyTorch reference
   - Tests broadcasting scenarios

### ⚠️ Issues to Address

1. **Half Dtype Not Implemented on Muxi Backend**
   - Multi-backend test shows: `FAIL: "cdist_cuda" not implemented for 'Half'`
   - This indicates the generic implementation doesn't support fp16
   - **Recommendation**: Add fp16 support or document limitation

2. **Hygon Backend Tuple Index Error**
   - Error: `tuple index out of range`
   - Likely related to shape handling in the kernel
   - **Recommendation**: Debug and fix for Hygon compatibility

3. **Only L2 Norm (p=2) Implemented**
   ```python
   def _cdist_forward(x1, x2, p=2.0, compute_mode=0):
       # Hardcoded p=2.0
   ```
   - PyTorch `cdist` supports arbitrary p values
   - Current implementation only handles L2 norm
   - **Recommendation**: Either:
     - Add support for other p values (p=1, p=inf, etc.)
     - Rename to `_cdist_forward_l2` to be explicit
     - Document the limitation clearly

4. **Missing Edge Case Tests**
   - Empty tensors (0 elements)
   - Single element tensors
   - Very large distances (potential overflow)
   - NaN/Inf in inputs

5. ** atol Relaxation in Commits**
   - Commit mentions "atol relaxation"
   - May indicate numerical precision issues
   - **Recommendation**: Verify numerical stability

### 🔍 Code Review Details

**Main Kernel** (`src/flag_gems/ops/_cdist_forward.py`):
```python
# Line ~20: Simple wrapper for generic case
def _cdist_forward(x1, x2, p=2.0, compute_mode=0):
    # Delegates to backend-specific implementation
```
- Good delegation pattern
- Should document p=2 limitation

**Metax Backend** (`src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py`):
```python
# Optimized kernel with tiling
# Uses shared memory for efficient distance computation
```
- Well-optimized for hardware
- 193 lines of specialized code

**Test File** (`tests/test_cdist_forward.py`):
```python
# Good test structure, but missing edge cases
```

### 📋 Test Suggestions

```python
def test_cdist_forward_empty():
    """Test with empty tensors"""
    x1 = torch.zeros(0, 3, device='cuda')
    x2 = torch.zeros(5, 3, device='cuda')
    result = flag_gems._cdist_forward(x1, x2)
    assert result.shape == (0, 5)

def test_cdist_forward_large_distance():
    """Test numerical stability with large values"""
    x1 = torch.tensor([[1e6, 0.0]], device='cuda')
    x2 = torch.tensor([[0.0, 0.0]], device='cuda')
    result = flag_gems._cdist_forward(x1, x2)
    # Should not overflow
    assert torch.isfinite(result).all()

def test_cdist_forward_half_dtype():
    """Test fp16 support"""
    x1 = torch.randn(10, 3, dtype=torch.float16, device='cuda')
    x2 = torch.randn(20, 3, dtype=torch.float16, device='cuda')
    result = flag_gems._cdist_forward(x1, x2)
    ref = torch.cdist(x1.float(), x2.float())
    # Check relative error is acceptable for fp16
    assert torch.allclose(result.float(), ref, rtol=1e-2, atol=1e-3)
```

### 📊 Multi-Backend Status

| Backend | Status | Speedup | Notes |
|---------|--------|---------|-------|
| Nvidia H20 | ✅ PASS | 2.437x | Primary target |
| Tianshu | ✅ PASS | - | Working |
| Muxi | ❌ FAIL | - | Half dtype issue |
| Ascend | ✅ PASS | - | Working |
| Hygon | ❌ FAIL | - | Tuple index error |

### 📊 Summary

| Category | Status |
|----------|--------|
| Implementation | ⚠️ L2-only limitation |
| Test Coverage | ⚠️ Missing edge cases |
| Backend Compatibility | ⚠️ Muxi/Hygon failures |
| Performance | ✅ 2.437x speedup |
| Documentation | ⚠️ Needs p=2 limitation docs |

**Overall**: Good implementation with excellent Metax optimization. Main concerns:
1. Fix Muxi Half dtype support
2. Fix Hygon tuple index error
3. Document L2-only limitation or add other p values
4. Add edge case tests

Ready to merge after addressing multi-backend failures.

**Review by Dongxu-H — COMMENTED** [review]
## Code Review: _cdist_forward Operator (Batched Pairwise Distance)

### 📋 Checklist Review

| # | Item | Status | Notes |
|---|------|--------|-------|
| 1 | CI Status | ✅ PASS | license/cla, triage passed |
| 2 | Merge Conflicts | ⚠️ CHECK | Verify mergeable state |
| 3 | Test Coverage | ✅ Good | Has `test_cdist_forward.py` |
| 4 | Benchmark Coverage | ✅ Good | Has benchmark with speedup metrics |
| 5 | Config Sort Order | ✅ Sorted | `_cdist_forward` in correct position |
| 6 | Kernel Registration | ✅ Correct | Registered in all `__init__.py` files |
| 7 | Missing Ops Check | ✅ N/A | No missing ops detected |
| 8 | NPU/CUDA Compatibility | ⚠️ Mixed | Metax backend specialized, Half dtype issue |
| 9 | Kernel Name vs Mark Name | ✅ Consistent | `_cdist_forward` used consistently |
| 10 | Logger Prefix | ✅ Good | Uses `flag_gems._cdist_forward` |
| 11 | Python Reserved Keywords | ✅ Good | No reserved keywords used |
| 12 | Code Quality | ✅ Good | Well-documented with comments |

### ✅ Strengths

1. **Complete Implementation Stack**
   - Main kernel: `src/flag_gems/ops/_cdist_forward.py`
   - Metax backend: `src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py`
   - Test file: `tests/test_cdist_forward.py`
   - Benchmark file: `benchmark/test_cdist_forward.py`
   - All configuration files updated

2. **Optimized Metax Backend**
   - 193 lines of specialized kernel for Metax hardware
   - Uses tiling strategy for efficient memory access
   - Achieves 2.437x speedup on Nvidia H20

3. **Good Algorithm Design**
   ```python
   # cdist(x1, x2, p=2) computes pairwise L2 distances
   # ||x1[i] - x2[j]||_2 for all i, j
   ```
   - Correctly implements L2 norm (p=2.0)
   - Handles batched inputs with broadcasting

4. **Comprehensive Tests**
   - Tests different shapes and dtypes
   - Compares with PyTorch reference
   - Tests broadcasting scenarios

### ⚠️ Issues to Address

1. **Half Dtype Not Implemented on Muxi Backend**
   - Multi-backend test shows: `FAIL: "cdist_cuda" not implemented for 'Half'`
   - This indicates the generic implementation doesn't support fp16
   - **Recommendation**: Add fp16 support or document limitation

2. **Hygon Backend Tuple Index Error**
   - Error: `tuple index out of range`
   - Likely related to shape handling in the kernel
   - **Recommendation**: Debug and fix for Hygon compatibility

3. **Only L2 Norm (p=2) Implemented**
   ```python
   def _cdist_forward(x1, x2, p=2.0, compute_mode=0):
       # Hardcoded p=2.0
   ```
   - PyTorch `cdist` supports arbitrary p values
   - Current implementation only handles L2 norm
   - **Recommendation**: Either:
     - Add support for other p values (p=1, p=inf, etc.)
     - Rename to `_cdist_forward_l2` to be explicit
     - Document the limitation clearly

4. **Missing Edge Case Tests**
   - Empty tensors (0 elements)
   - Single element tensors
   - Very large distances (potential overflow)
   - NaN/Inf in inputs

5. ** atol Relaxation in Commits**
   - Commit mentions "atol relaxation"
   - May indicate numerical precision issues
   - **Recommendation**: Verify numerical stability

### 🔍 Code Review Details

**Main Kernel** (`src/flag_gems/ops/_cdist_forward.py`):
```python
# Line ~20: Simple wrapper for generic case
def _cdist_forward(x1, x2, p=2.0, compute_mode=0):
    # Delegates to backend-specific implementation
```
- Good delegation pattern
- Should document p=2 limitation

**Metax Backend** (`src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py`):
```python
# Optimized kernel with tiling
# Uses shared memory for efficient distance computation
```
- Well-optimized for hardware
- 193 lines of specialized code

**Test File** (`tests/test_cdist_forward.py`):
```python
# Good test structure, but missing edge cases
```

### 📋 Test Suggestions

```python
def test_cdist_forward_empty():
    """Test with empty tensors"""
    x1 = torch.zeros(0, 3, device='cuda')
    x2 = torch.zeros(5, 3, device='cuda')
    result = flag_gems._cdist_forward(x1, x2)
    assert result.shape == (0, 5)

def test_cdist_forward_large_distance():
    """Test numerical stability with large values"""
    x1 = torch.tensor([[1e6, 0.0]], device='cuda')
    x2 = torch.tensor([[0.0, 0.0]], device='cuda')
    result = flag_gems._cdist_forward(x1, x2)
    # Should not overflow
    assert torch.isfinite(result).all()

def test_cdist_forward_half_dtype():
    """Test fp16 support"""
    x1 = torch.randn(10, 3, dtype=torch.float16, device='cuda')
    x2 = torch.randn(20, 3, dtype=torch.float16, device='cuda')
    result = flag_gems._cdist_forward(x1, x2)
    ref = torch.cdist(x1.float(), x2.float())
    # Check relative error is acceptable for fp16
    assert torch.allclose(result.float(), ref, rtol=1e-2, atol=1e-3)
```

### 📊 Multi-Backend Status

| Backend | Status | Speedup | Notes |
|---------|--------|---------|-------|
| Nvidia H20 | ✅ PASS | 2.437x | Primary target |
| Tianshu | ✅ PASS | - | Working |
| Muxi | ❌ FAIL | - | Half dtype issue |
| Ascend | ✅ PASS | - | Working |
| Hygon | ❌ FAIL | - | Tuple index error |

### 📊 Summary

| Category | Status |
|----------|--------|
| Implementation | ⚠️ L2-only limitation |
| Test Coverage | ⚠️ Missing edge cases |
| Backend Compatibility | ⚠️ Muxi/Hygon failures |
| Performance | ✅ 2.437x speedup |
| Documentation | ⚠️ Needs p=2 limitation docs |

**Overall**: Good implementation with excellent Metax optimization. Main concerns:
1. Fix Muxi Half dtype support
2. Fix Hygon tuple index error
3. Document L2-only limitation or add other p values
4. Add edge case tests

Ready to merge after addressing multi-backend failures.

**Review by Dongxu-H — DISMISSED** [review]
The June 18 review flagged several concerns which have been addressed in subsequent commits. The kernel was moved from Metax backend to generic ops module in the June 23 commit, making it available for all backends. The import ordering was fixed in the June 26 isort fix commit.

Implementation correctness: The kernel correctly computes L2 pairwise distances using the formula sqrt of sum of squared differences. The BLOCK_M parameter is dynamically chosen as min of next_power_of_2 of M and 64 for efficient memory access. The kernel handles batch dimensions with proper stride calculations and supports broadcasting between input tensors.

p value handling: The implementation correctly falls back to torch.cdist for p values other than 2.0, which handles the limitation gracefully. The assertion for float32 only is documented and consistent with the test coverage.

Configuration ordering: The operators.yaml entry for cdist_forward is correctly placed between cdist_backward at line 1264 and ceil at line 1277. The __init__.py registration places _cdist_forward between _cdist_backward and _conv_depthwise2d which is alphabetically correct. The ops/__init__.py import ordering was fixed in the June 26 commit to follow isort conventions.

File: benchmark/test_cdist_forward.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            x2 = torch.randn(*shape2, dtype=cur_dtype, device=self.device)
+            yield x1, x2, 2.0
+
+    def get_tflops(self, op, *args, **kwargs):
+        x1, x2, _ = args
+        # FLOPs = 2 * P * R * M (for L2 distance computation)
+        return 2 * x1.shape[-2] * x2.shape[-2] * x1.shape[-1]
+
+
+@pytest.mark.cdist_forward
```
Comment (tengqm) [inline benchmark/test_cdist_forward.py:35]: why don't just call this 'cdist'?

File: src/flag_gems/ops/_cdist_forward.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            BLOCK_M=BLOCK_M,
+        )
+
+    # Reshape output to broadcast batch shape
+    output = output.view(*batch_out, P, R)
+
+    return output
+
+
+__all__ = ["_cdist_forward"]
```
Comment (tengqm) [inline src/flag_gems/ops/_cdist_forward.py]: Remove this line.

File: src/flag_gems/ops/cdist.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
+# Copyright 2026, The FlagOS Contributors.
```
Comment (tengqm) [inline src/flag_gems/ops/cdist.py:1]: please consider merge the code with the backward implementation

File: conf/operators.yaml [hunk 4]
```diff
@@ -1274,6 +1274,20 @@ ops:
       - Math
     stages:
       - alpha: '5.4'
+  - id: cdist_forward
```
Comment (tengqm) [inline conf/operators.yaml:1532]: for most operators, the "forward" case doesn't need a 'forward' affix.

File: src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2026, The FlagOS Contributors.
```
Comment (tengqm) [inline src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py:1]: Remove this garbage file.

File: src/flag_gems/runtime/backend/_metax/ops/__init__.py [hunk 6]
```diff
@@ -35,6 +36,7 @@
 from .zeros_like import zeros_like
 
 __all__ = [
+    "_cdist_forward",
```
Comment (tengqm) [inline src/flag_gems/runtime/backend/_metax/ops/__init__.py]: Remove this line.

File: benchmark/test_cdist_forward.py (hunk 1 above)
Comment (Dongxu-H) [reply benchmark/test_cdist_forward.py:35]: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L4526

We should follow the same aten function name

- func: _cdist_forward(Tensor x1, Tensor x2, float p, int? compute_mode) -> Tensor
  dispatch:
    CPU, CUDA, MPS, XPU: _cdist_forward
    MTIA: _cdist_forward_mtia
  autogen: _cdist_forward.out
  tags: core

File: conf/operators.yaml (hunk 4 above)
Comment (Dongxu-H) [reply conf/operators.yaml:1532]: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L4526

We should follow the same aten function name

- func: _cdist_forward(Tensor x1, Tensor x2, float p, int? compute_mode) -> Tensor
  dispatch:
    CPU, CUDA, MPS, XPU: _cdist_forward
    MTIA: _cdist_forward_mtia
  autogen: _cdist_forward.out
  tags: core

File: src/flag_gems/ops/cdist.py (hunk 3 above)
Comment (kkkwb) [reply src/flag_gems/ops/cdist.py:1]: Done as suggested. The two files [_cdist_forward.py] and [cdist_backward.py] are merged into [cdist.py], with both kernels and wrappers preserved verbatim in the same file.Commit 8c034fd64742a07fb91b16c389e6ef0919874db7.

File: src/flag_gems/ops/_cdist_forward.py (hunk 2 above)
Comment (kkkwb) [reply src/flag_gems/ops/_cdist_forward.py]: Done. Dropped during the merge; not retained in [cdist.py]. Commit 8c034fd64742a07fb91b16c389e6ef0919874db7.

File: src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py (hunk 5 above)
Comment (kkkwb) [reply src/flag_gems/runtime/backend/_metax/ops/_cdist_forward.py:1]: The file is deleted. Commit 8c034fd64742a07fb91b16c389e6ef0919874db7.

File: src/flag_gems/runtime/backend/_metax/ops/__init__.py (hunk 6 above)
Comment (kkkwb) [reply src/flag_gems/runtime/backend/_metax/ops/__init__.py]: Done. Removed together with the line-1 import from ._cdist_forward import _cdist_forward, so that deleting the stub does not leave a failing import that would break the Metax ops package loading. The Metax label and related description in the cdist_forward entry of [conf/operators.yaml] are also removed accordingly. Commit 8c034fd64742a07fb91b16c389e6ef0919874db7.

**Comment by Caeruleann** [comment]
Please resolve conflicts.

**Review by bin913 — APPROVED** [review]
